infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -126,80 +100,120 @@ public void delete(@Nonnull ConflictTxStream key, Long globalAddress, @Nonnull R
+        if (cacheEntries.isEmpty()) {
+            return Address.NOT_FOUND;
+        }
+        return cacheEntries.peek().txVersion;
+    }
 
-        conflictCache.asMap().forEach((key, txVersion) -> {
-            if (txVersion == null || txVersion >= trimMark) {
-                return;
-            }
+    /**
+     * Invalidate the records with the minAddress. It could be one or multiple records
+     * @return the number of entries has been invalidated and removed from the cache.
+     */
+    private int invalidateFirst() {
+        ConflictTxStream firstEntry = cacheEntries.peek();
+        if (firstEntry == null) {
+            return 0;
+        }
 
-            conflictCache.invalidate(key);
-            entries.incrementAndGet();
-        });
+        int numEntries = 0;
+        while (firstAddress() == firstEntry.txVersion) {
+            ConflictTxStream entry = cacheEntries.poll();
+            cacheConflictKeys.remove(entry);
+            numEntries ++;
+        }
 
-        log.info("Invalidated entries: {}", entries.get());
+        log.trace("Evict {} entries", numEntries);
+        maxConflictWildcard = Math.max(maxConflictWildcard, firstEntry.txVersion);
+        return numEntries;
     }
 
     /**
-     * The cache size
+     * Invalidate all records up to a trim mark (not included).
      *
-     * @return cache size
+     * @param trimMark trim mark
      */
-    public long size() {
-        return conflictCache.estimatedSize();
+    public void invalidateUpTo(long trimMark) {
+        log.debug("Invalidate sequencer cache. Trim mark: {}", trimMark);
+        int entries = 0;
+        int pqEntries = 0;
+        while (Address.isAddress(firstAddress()) && firstAddress() < trimMark) {
+            pqEntries += invalidateFirst();
+            entries++;
+        }
+        log.info("Invalidated entries {} addresses {}", pqEntries, entries);
     }
 
     /**
-     * Put a value in the cache
+     * The cache size as the number of entries
      *
-     * @param conflictStream conflict stream
-     * @param newTail        global tail
+     * @return cache size
      */
-    public void put(ConflictTxStream conflictStream, long newTail) {
-        conflictCache.put(conflictStream, newTail);
+    public int size() {
+        return cacheConflictKeys.size();
     }
 
     /**
-     * Discard all entries in the cache
+     * The memory space used by the entries and also the space used by
+     * priority queue and hashmap to store the pointers
+     * @return the memory space used in bytes:
      */
-    public void invalidateAll() {
-        log.info("Invalidate sequencer server cache");
-        conflictCache.invalidateAll();
+    public long byteSize() {
+        log.debug("the cache has {} entries,  the object size used {}, calculated by beepSize {}",
+                size(), ConflictTxStream_Obj_Size,
+                cacheEntries.isEmpty() ? 0 : MetricsUtils.sizeOf.deepSizeOf(cacheEntries.peek()));
```

Comment:
You can't use `deepSizeOf ` so often, it's not desired operation. Avoid it as much as possible, only in very rare and limited situations when there is no choice then you can use it, otherwise not.

Reply:
This is the debug mode. Also, deepSize is costly only when you pass in an array like structure and the object has embeded deep pointers.